packages/cli/src/commonlib/codeFlowTenantLogin.ts

```diff
@@ -0,0 +1,329 @@
+      prompt: "select_account"
+    };
+
+    let deferredRedirect: Deferred<string>;
+    const redirectPromise: Promise<string> = new Promise<string>(
+      (resolve, reject) => (deferredRedirect = { resolve, reject })
+    );
+
+    app.get("/", (req: express.Request, res: express.Response) => {
+      const tokenRequest = {
+        code: req.query.code as string,
+        scopes: this.scopes!,
+        redirectUri: `http://localhost:${serverPort}`,
+        codeVerifier: codeVerifier
+      };
+
+      this.pca!.acquireTokenByCode(tokenRequest)
+        .then(async (response) => {
+          if (response) {
+            if (response.account) {
+              await this.mutex?.runExclusive(async () => {
+                this.account = response.account!;
+              });
+              deferredRedirect.resolve(response.accessToken);
+
+              sendFile(
+                res,
+                path.join(__dirname, "./codeFlowResult/index.html"),
+                "text/html; charset=utf-8",
+                this.accountName!
+              );
+            }
+          } else {
+            throw new Error("get no response");
+          }
+        })
+        .catch((error) => {
+          CliCodeLogInstance.error("[Login] " + error.message);
+          deferredRedirect.reject(error);
+          res.status(500).send(error);
+        });
+    });
+
+    const codeTimer = setTimeout(() => {
+      deferredRedirect.reject(
+        returnSystemError(
+          new Error(ErrorMessage.timeoutMessage),
+          ErrorMessage.component,
+          ErrorMessage.loginError
+        )
+      );
+    }, 5 * 60 * 1000);
+
+    function cancelCodeTimer() {
+      clearTimeout(codeTimer);
+    }
+
+    let accessToken = undefined;
+    try {
+      await this.startServer(server, serverPort!);
+      this.pca!.getAuthCodeUrl(authCodeUrlParameters).then(async (response: string) => {
+        // TODO change console.log to logProvider, for now, logProvider may be hidden
+        if (this.accountName == "azure") {
+          console.log(colors.green(`[${constants.cliSource}] ${azureLoginMessage}`));
+        } else {
+          console.log(colors.green(`[${constants.cliSource}] ${m365LoginMessage}`));
+        }
+        open(response);
+      });
+
+      redirectPromise.then(cancelCodeTimer, cancelCodeTimer);
+      accessToken = await redirectPromise;
+    } finally {
+      server.close();
+    }
+
+    return accessToken;
+  }
+
+  async logout(): Promise<boolean> {
+    const accountCache = String(fs.readFileSync(accountPath + this.accountName, UTF8));
+    const dataCache = await this.msalTokenCache!.getAccountByHomeId(accountCache);
+    if (dataCache) {
+      this.msalTokenCache?.removeAccount(dataCache);
+    }
+    if (fs.existsSync(accountPath + this.accountName)) {
+      fs.writeFileSync(accountPath + this.accountName, "", UTF8);
+    }
+    return true;
+  }
+
+  async getToken(tenantId?: string): Promise<string | undefined> {
```

Comment:
Duplicated `getToken` and `getTenantToken`. This work flow should only work for tenanted login.

Reply:
```getTenantToken``` has been removed.